base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -157,9 +159,12 @@ pub trait BlockchainBackend: Send + Sync {
     ) -> Result<HashOutput, ChainStorageError>;
     /// Constructs a merkle proof for the specified merkle mountain range and the given leaf position.
     fn fetch_mmr_proof(&self, tree: MmrTree, pos: usize) -> Result<MerkleProof, ChainStorageError>;
-    /// Fetches the checkpoint corresponding to the provided height, the checkpoint consist of the list of nodes
-    /// added & deleted for the given Merkle tree.
+    /// Fetches the `MerkleCheckPoint` corresponding to the given height. In pruned mode, the underlying database may
+    /// not be able to provide the checkpoint as it has been merged into the base checkpoint. In this case a
+    /// `BeyondPruningHorizon` error is returned.
     fn fetch_checkpoint(&self, tree: MmrTree, height: u64) -> Result<MerkleCheckPoint, ChainStorageError>;
+    /// Fetches the `MerkleCheckPoint` at the given index
+    fn fetch_checkpoint_at(&self, tree: MmrTree, index: usize) -> Result<Option<MerkleCheckPoint>, ChainStorageError>;
```

Comment:
Whats the diff here. 
Would checkpoint at always reference the first checkpoint the node has, while height is the chain height?

Reply:
Yup exactly, this fetches by index because the "temporary" checkpoints don't correspond to any particular height.

Reply:
It might be best to rename `fetch_checkpoint` to `fetch_checkpoint_at_height`

Reply:
Agreed - done